walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -18,18 +18,138 @@
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
-import java.util.Map;
 import java.util.Set;
 
 public final class FindMeetingQuery {
+  /**
+   * Return the TimeRanges available for all meeting members given existing
+   * events
+   */
+  public Collection<TimeRange> query(Collection<Event> events,
+                                     MeetingRequest request) {
+    // For a meeting longer than a day, no times are possible
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // Get collections of people from MeetingRequest
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    // With no attendees, the meeting can be held at any time
+    if (attendees.isEmpty() && optionalAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    long duration = request.getDuration();
+
+    // If either group is empty, only need to consider the other
+    if (attendees.isEmpty() || optionalAttendees.isEmpty()) {
+      Collection<String> nonEmptyAttendees =
+          attendees.isEmpty() ? optionalAttendees : attendees;
+      return combineAvailability(events, nonEmptyAttendees, duration);
+    }
+
+    ArrayList<TimeRange> mandatoryAvailability =
+        combineAvailability(events, attendees, duration);
+
+    // If there are no open slots between mandatory attendees,
+    // no need to consider optional attendees
+    if (mandatoryAvailability.isEmpty()) {
+      return mandatoryAvailability;
+    }
+
+    // Get individual availabilities for optional attendees
+    ArrayList<ArrayList<TimeRange>> optionalAttendeesCals =
+        new ArrayList<ArrayList<TimeRange>>();
+        
+    for (String curPerson : optionalAttendees) {
+      optionalAttendeesCals.add(
+          combineAvailability(events, Arrays.asList(curPerson), duration));
+    }
+
+    // Meetings can be held every {increment} minutes
+    // E.g. if all meetings are only held at 15 minute increments of the hour
+    // for durations that are multiples of 15 minutes, can reduce amount of
+    // computation by setting {increment} to 15
+    int increment = 1;
+
+    // For all possible times every {increment} minutes that allows for a 
+    // meeting of the desired duration, store the number of optional attendees 
+    // who can attend
+    ArrayList<ArrayList<Integer>> forEachDuration =
+        new ArrayList<ArrayList<Integer>>();
```

Comment:
Using the name forEachDuration as well as "meeting of the desired duration" also caused confusion.  This isn't the meeting duration but rather the potential start time.